File: haskell/lib/D_fut.hs
module Futures
  ( DFut(..)
  , new
  , share
  ) where

import qualified DStoreId

data DFut a = DFut
  { dStoreId :: DStoreId.DStoreId
  }

new :: DStoreId.DStoreId -> DFut a
new dStoreId = DFut { dStoreId = dStoreId }

share :: DFut a -> DFut a
share dfut = DFut { dStoreId = DStoreId.clone (dStoreId dfut) }

instance DStoreId.FromDStoreId (DFut a) where
  fromDStoreId = new

File: go/d_fut.go
package futures

import (
    "futures/dstore"
)

type DFut[T any] struct {
    DStoreID dstore.DStoreId
}

func New[T any](dStoreID dstore.DStoreId) DFut[T] {
    return DFut[T]{
        DStoreID: dStoreID,
    }
}

func (d DFut[T]) Share() DFut[T] {
    return DFut[T]{
        DStoreID: d.DStoreID.Clone(),
    }
}

func FromDStoreId[T any](dStoreID dstore.DStoreId) DFut[T] {
    return New[T](dStoreID)
}